zcash_keys/src/keys/zcashd.rs

```diff
@@ -0,0 +1,245 @@
+    const ZCASHD_LEGACY_ACCOUNT_INDEX: u32 = 0x7FFFFFFF;
+
+    // Parses a HD key path to obtain the coin type and account ID or legacy Sapling address index.
+    //
+    // Historically, each key generated via the `z_getnewaddress` RPC method was treated as a separate
+    // pool of funds, much like a ZIP 32 account. After the introduction of BIP 39 mnemonic seed
+    // phrases in Zcashd v4.7.0, we wanted to retain these semantics but did not want `z_getnewaddress`
+    // to implicitly create a new ZIP 32 account. Instead, we used a fixed keypath scheme of
+    //
+    // `m/32'/coin_type'/0x7FFFFFFF'/addressIndex'`
+    //
+    // for each new address. This is not a "standard" path, but instead is a predictable location for
+    // legacy zcashd-derived keys that is minimally different from the UA account path, while unlikely
+    // to collide with normal UA account usage.
+    //
+    // For post v4.7.0 Zcashd accounts, we use standard ZIP 32 HD derivation from the master key to
+    // obtain the unified spending key.
+    pub fn parse_hd_path<C: NetworkConstants>(
+        network: &C,
+        path: &str,
+    ) -> Result<Self, PathParseError> {
+        let re = Regex::new(r"^m/32'/(\d+)'/(\d+)'(?:/(\d+)')?$").expect("checked to be valid");
+        let parts = re.captures(path).ok_or(PathParseError::PathInvalid)?;
+        assert!(parts.len() <= 4);
+
+        let coin_type = parts
+            .get(1)
+            .map_or(Err(PathParseError::PathInvalid), |part| {
+                let part_str = part.as_str();
+                part_str
+                    .parse::<u32>()
+                    .map_err(|_| PathParseError::CoinTypeInvalid(part_str.to_string()))
+            })?;
+        if coin_type != network.coin_type() {
+            return Err(PathParseError::CoinTypeMismatch {
+                expected: network.coin_type(),
+                actual: coin_type,
+            });
+        }
+
+        let account_index = parts
+            .get(2)
+            .map_or(Err(PathParseError::PathInvalid), |part| {
+                let part_str = part.as_str();
+                part_str
+                    .parse::<u32>()
+                    .map_err(|_| PathParseError::AccountIndexInvalid(part_str.to_string()))
+            })?;
+
+        if account_index == Self::ZCASHD_LEGACY_ACCOUNT_INDEX {
+            let address_index =
+                parts
+                    .get(3)
+                    .map_or(Err(PathParseError::AddressIndexMissing), |part| {
+                        let part_str = part.as_str();
+
+                        match part_str.parse::<u32>() {
+                            Ok(v) if v < (1 << 31) => Ok(LegacyAddressIndex(v)),
+                            _ => Err(PathParseError::AddressIndexInvalid(part_str.to_string())),
+                        }
+                    })?;
+
+            Ok(ZcashdHdDerivation::Post470LegacySapling { address_index })
+        } else {
+            if parts.get(4).is_some() {
```

Comment:
Bug:
```suggestion
            if parts.get(3).is_some() {
```

This was introduced by 5b879729ced414c56d61342c8cb4f4999f2a1391 which was pushed directly to the `zallet_dev` branch. The fix to this must also be made to that branch, not just to this PR.